### PR TITLE
Azure blockblob backend parametrized connection/read timeouts

### PR DESCRIPTION
**Pull request by Tomasz-Kluczkowski**
## Description

Introduce connection and read timeout settings for Azure blockblob backend.

### Motivation
We are using celery in production with Azure blockblob backend. 

About 2-3 times a day we noticed that the tasks would soft/hard time out for no reason. I have carried an investigation and managed to obtain stack traces (many of them, all stopping at same line) of the executing task at the moment of it receiving SIGUSR1/SIGTERM from the supervising process (used signals for that).

what I noticed, when celery was trying to call `store_result` and then inside of it `_get_task_meta_for` it would hang on an ssl socket read.

This is what the typical stack trace looks like:

```python
File "/usr/local/lib/python3.9/site-packages/celery/backends/base.py", line 482, in store_result
    self._store_result(task_id, result, state, traceback,
  File "/usr/local/lib/python3.9/site-packages/celery/backends/base.py", line 903, in _store_result
    current_meta = self._get_task_meta_for(task_id)
  File "/usr/local/lib/python3.9/site-packages/celery/backends/base.py", line 925, in _get_task_meta_for
    meta = self.get(self.get_key_for_task(task_id))
  File "/usr/local/lib/python3.9/site-packages/celery/backends/azureblockblob.py", line 91, in get
    return blob_client.download_blob().readall().decode()
  File "/usr/local/lib/python3.9/site-packages/azure/core/tracing/decorator.py", line 83, in wrapper_use_tracer
    return func(*args, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/azure/storage/blob/_blob_client.py", line 828, in download_blob
    return StorageStreamDownloader(**options)
  File "/usr/local/lib/python3.9/site-packages/azure/storage/blob/_download.py", line 349, in __init__
    self._response = self._initial_request()
  File "/usr/local/lib/python3.9/site-packages/azure/storage/blob/_download.py", line 386, in _initial_request
    location_mode, response = self._clients.blob.download(
  File "/usr/local/lib/python3.9/site-packages/azure/storage/blob/_generated/operations/_blob_operations.py", line 179, in download
    pipeline_response = self._client._pipeline.run(request, stream=True, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/azure/core/pipeline/_base.py", line 211, in run
    return first_node.send(pipeline_request)  # type: ignore
  File "/usr/local/lib/python3.9/site-packages/azure/core/pipeline/_base.py", line 71, in send
    response = self.next.send(request)
  File "/usr/local/lib/python3.9/site-packages/azure/core/pipeline/_base.py", line 71, in send
    response = self.next.send(request)
  File "/usr/local/lib/python3.9/site-packages/azure/core/pipeline/_base.py", line 71, in send
    response = self.next.send(request)
  [Previous line repeated 2 more times]
  File "/usr/local/lib/python3.9/site-packages/azure/core/pipeline/policies/_redirect.py", line 158, in send
    response = self.next.send(request)
  File "/usr/local/lib/python3.9/site-packages/azure/core/pipeline/_base.py", line 71, in send
    response = self.next.send(request)
  File "/usr/local/lib/python3.9/site-packages/azure/storage/blob/_shared/policies.py", line 489, in send
    response = self.next.send(request)
  File "/usr/local/lib/python3.9/site-packages/azure/core/pipeline/_base.py", line 71, in send
    response = self.next.send(request)
  File "/usr/local/lib/python3.9/site-packages/azure/core/pipeline/_base.py", line 71, in send
    response = self.next.send(request)
  File "/usr/local/lib/python3.9/site-packages/azure/core/pipeline/_base.py", line 71, in send
    response = self.next.send(request)
  [Previous line repeated 1 more time]
  File "/usr/local/lib/python3.9/site-packages/azure/storage/blob/_shared/policies.py", line 290, in send
    response = self.next.send(request)
  File "/usr/local/lib/python3.9/site-packages/azure/core/pipeline/_base.py", line 71, in send
    response = self.next.send(request)
  File "/usr/local/lib/python3.9/site-packages/azure/core/pipeline/_base.py", line 71, in send
    response = self.next.send(request)
  File "/usr/local/lib/python3.9/site-packages/azure/core/pipeline/_base.py", line 103, in send
    self._sender.send(request.http_request, **request.context.options),
  File "/usr/local/lib/python3.9/site-packages/azure/storage/blob/_shared/base_client.py", line 333, in send
    return self._transport.send(request, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/azure/core/pipeline/transport/_requests_basic.py", line 273, in send
    response = self.session.request(  # type: ignore
  File "/usr/local/lib/python3.9/site-packages/requests/sessions.py", line 542, in request
    resp = self.send(prep, **send_kwargs)
  File "/usr/local/lib/python3.9/site-packages/requests/sessions.py", line 655, in send
    r = adapter.send(request, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/requests/adapters.py", line 439, in send
    resp = conn.urlopen(
  File "/usr/local/lib/python3.9/site-packages/urllib3/connectionpool.py", line 699, in urlopen
    httplib_response = self._make_request(
  File "/usr/local/lib/python3.9/site-packages/urllib3/connectionpool.py", line 440, in _make_request
    httplib_response = conn.getresponse()
  File "/usr/local/lib/python3.9/http/client.py", line 1347, in getresponse
    response.begin()
  File "/usr/local/lib/python3.9/http/client.py", line 307, in begin
    version, status, reason = self._read_status()
  File "/usr/local/lib/python3.9/http/client.py", line 268, in _read_status
    line = str(self.fp.readline(_MAXLINE + 1), "iso-8859-1")
  File "/usr/local/lib/python3.9/socket.py", line 704, in readinto
    return self._sock.recv_into(b)
  File "/usr/local/lib/python3.9/ssl.py", line 1241, in recv_into
    return self.read(nbytes, buffer)
  File "/usr/local/lib/python3.9/ssl.py", line 1099, in read
    return self._sslobj.read(len, buffer)
```

I looked why in debug mode and the key takeaway is since celery backend blockblob does not specify the timeouts, the azure blob library will use its defaults defined in `python3.9/site-packages/azure/storage/blob/_shared/constants.py`

and the main culprit there is this:

```python
# for python 3.5+, there was a change to the definition of the socket timeout (as far as socket.sendall is concerned)
# The socket timeout is now the maximum total duration to send all data.
if sys.version_info >= (3, 5):
    # the timeout to connect is 20 seconds, and the read timeout is 80000 seconds
    # the 80000 seconds was calculated with:
    # 4000MB (max block size)/ 50KB/s (an arbitrarily chosen minimum upload speed)
    READ_TIMEOUT = 80000
```

So we get 80000s (55.555[...] days !!!!) to timeout on reads.....great :). It is pretty much same as infinite. 
Screenshot 2021-09-24 at 13 09 06

With the changes proposed we keep the default connection timeout at 20s and set the read timeout to 120s which should be sufficient for most day-to-day usage in older and newer python versions. 

If we want to make similar distinction between read timeout values for older and newer python versions as in `azure.storage.blob._shared.constants` - please let me know. Currently this makes it 120s for both, whereas azure has 20s for older python versions (there it is measured to first packet read, instead of whole read time so can be shorter).

If we think the default read timeout should be higher just in case - please let me know about what value you think is sufficient. My assumption is, the backend results are small and are transferred over same network as other services in the application in production use, therefore bandwidth should not be an issue (we use 5s for this settings in our production without any issues).

Finally, we have tested this improvement in our production and the timeouts disappeared :).

### CODE REVIEW

**Review by auvipy — CHANGES_REQUESTED** [review]
thanks for the enhancement. can you also add unit tests for this?

**Comment by Tomasz-Kluczkowski** [comment]
Looks like I forgot to add little test confirming the client is configured correctly - will do hopefully over this weekend.

File: docs/userguide/configuration.rst [hunk 1]
```diff
@@ -1599,6 +1599,24 @@ Default: 3.
...
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Default: 20.
+
+Timeout in seconds for establishing the azure block blob connection.
+
+.. setting:: azureblockblob_read_timeout
+
+``azureblockblob_read_timeout``
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
Comment (thedrow) [inline docs/userguide/configuration.rst:1614]:
```suggestion
``azureblockblob_read_timeout``
~~~~~~~~~~~~~~~~~~~~
```

**Review by thedrow — CHANGES_REQUESTED** [review]

**Comment by auvipy** [comment]
should we consider this for 5.2?

**Comment by thedrow** [comment]
> 
> 
> should we consider this for 5.2?

It's a small feature so we might.

**Review by auvipy — APPROVED** [review]

**Pull request merged**